{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assignment 1: Search for SNP and deletions in a sample\n",
    "\n",
    "In order to obtain biological information on SNPs, insertions and deletions, manipulate SAM files in the following way:\n",
    "1. Convert SAM file obtained in LAB2 to BAM\n",
    "2. Sort BAM to make the following process faster.\n",
    "3. Use bcftools to obtain a VCF file\n",
    "4. Write a Python program to parse the VCF file and obtain only Single Nucleotides Polymorphism (SNP) for which the information is complete.\n",
    "5. Write a Python program to parse VCF file and obtain only Insertions or Deletions (INDEL) for which the information is complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>ID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "      <th>Format</th>\n",
       "      <th>bam file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>48636</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=1;I16=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0;QS=0,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>48637</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0;QS=0,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>48638</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,2,4,0...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>48639</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,3,9,0...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>48640</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,4,16,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10</td>\n",
       "      <td>48641</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,5,25,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10</td>\n",
       "      <td>48642</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[T, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,1,0,0,0,70,4900,0,0,60,3600,0,0,6...</td>\n",
       "      <td>PL</td>\n",
       "      <td>60,3,0,60,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10</td>\n",
       "      <td>48643</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,7,49,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10</td>\n",
       "      <td>48644</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,8,64,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>48645</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,9,81,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>48646</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,10,10...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>48647</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,11,12...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>10</td>\n",
       "      <td>48648</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,12,14...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>48649</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,77,5929,0,0,60,3600,0,0,13,16...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>10</td>\n",
       "      <td>48650</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>[&lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=1,0,0,0,73,5329,0,0,60,3600,0,0,14,19...</td>\n",
       "      <td>PL</td>\n",
       "      <td>0,3,60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Position ID Reference Alternative Quality Filter  \\\n",
       "0          10    48636  .         T       [<*>]       0      .   \n",
       "1          10    48637  .         G       [<*>]       0      .   \n",
       "2          10    48638  .         A       [<*>]       0      .   \n",
       "3          10    48639  .         A       [<*>]       0      .   \n",
       "4          10    48640  .         G       [<*>]       0      .   \n",
       "5          10    48641  .         A       [<*>]       0      .   \n",
       "6          10    48642  .         C    [T, <*>]       0      .   \n",
       "7          10    48643  .         C       [<*>]       0      .   \n",
       "8          10    48644  .         T       [<*>]       0      .   \n",
       "9          10    48645  .         G       [<*>]       0      .   \n",
       "10         10    48646  .         C       [<*>]       0      .   \n",
       "11         10    48647  .         C       [<*>]       0      .   \n",
       "12         10    48648  .         C       [<*>]       0      .   \n",
       "13         10    48649  .         G       [<*>]       0      .   \n",
       "14         10    48650  .         A       [<*>]       0      .   \n",
       "\n",
       "                                                 Info Format        bam file  \n",
       "0   DP=1;I16=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0;QS=0,...     PL           0,0,0  \n",
       "1   DP=2;I16=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0;QS=0,...     PL           0,0,0  \n",
       "2   DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,2,4,0...     PL          0,3,60  \n",
       "3   DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,3,9,0...     PL          0,3,60  \n",
       "4   DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,4,16,...     PL          0,3,60  \n",
       "5   DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,5,25,...     PL          0,3,60  \n",
       "6   DP=2;I16=0,0,1,0,0,0,70,4900,0,0,60,3600,0,0,6...     PL  60,3,0,60,3,60  \n",
       "7   DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,7,49,...     PL          0,3,60  \n",
       "8   DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,8,64,...     PL          0,3,60  \n",
       "9   DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,9,81,...     PL          0,3,60  \n",
       "10  DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,10,10...     PL          0,3,60  \n",
       "11  DP=2;I16=1,0,0,0,74,5476,0,0,60,3600,0,0,11,12...     PL          0,3,60  \n",
       "12  DP=2;I16=1,0,0,0,70,4900,0,0,60,3600,0,0,12,14...     PL          0,3,60  \n",
       "13  DP=2;I16=1,0,0,0,77,5929,0,0,60,3600,0,0,13,16...     PL          0,3,60  \n",
       "14  DP=2;I16=1,0,0,0,73,5329,0,0,60,3600,0,0,14,19...     PL          0,3,60  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCF_FILE = \"sorted.vcf\"\n",
    "\n",
    "file = parse_vcf(VCF_FILE)\n",
    "\n",
    "df = pd.DataFrame(file, columns=['Chromosome', 'Position', 'ID', 'Reference', 'Alternative', 'Quality', 'Filter', 'Info', 'Format', 'bam file'])\n",
    "df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>ID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "      <th>Format</th>\n",
       "      <th>bam file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10</td>\n",
       "      <td>48642</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[T, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,1,0,0,0,70,4900,0,0,60,3600,0,0,6...</td>\n",
       "      <td>PL</td>\n",
       "      <td>60,3,0,60,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>10</td>\n",
       "      <td>48667</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>[G, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,1,0,0,0,77,5929,0,0,60,3600,0,0,2...</td>\n",
       "      <td>PL</td>\n",
       "      <td>60,3,0,60,3,60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>10</td>\n",
       "      <td>3642986</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>[C, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,1,0,0,0,77,5929,0,0,7,49,0,0,25,6...</td>\n",
       "      <td>PL</td>\n",
       "      <td>7,3,0,7,3,7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>10</td>\n",
       "      <td>3643007</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[C, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,1,0,0,0,70,4900,0,0,7,49,0,0,25,6...</td>\n",
       "      <td>PL</td>\n",
       "      <td>7,3,0,7,3,7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>10</td>\n",
       "      <td>19491103</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>[G, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=1;I16=0,0,1,0,0,0,40,1600,0,0,60,3600,0,0,1...</td>\n",
       "      <td>PL</td>\n",
       "      <td>40,3,0,40,3,40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6960</th>\n",
       "      <td>18</td>\n",
       "      <td>78994238</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[T, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,2...</td>\n",
       "      <td>PL</td>\n",
       "      <td>40,3,0,40,3,40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6963</th>\n",
       "      <td>18</td>\n",
       "      <td>78994241</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[T, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,2...</td>\n",
       "      <td>PL</td>\n",
       "      <td>40,3,0,40,3,40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6970</th>\n",
       "      <td>18</td>\n",
       "      <td>78994248</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,1...</td>\n",
       "      <td>PL</td>\n",
       "      <td>40,3,0,40,3,40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6972</th>\n",
       "      <td>18</td>\n",
       "      <td>78994250</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>[G, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,1...</td>\n",
       "      <td>PL</td>\n",
       "      <td>40,3,0,40,3,40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7002</th>\n",
       "      <td>18</td>\n",
       "      <td>79891262</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>[C, &lt;*&gt;]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>DP=2;I16=0,0,1,0,0,0,80,6400,0,0,28,784,0,0,25...</td>\n",
       "      <td>PL</td>\n",
       "      <td>28,3,0,28,3,28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>159 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome  Position ID Reference Alternative Quality Filter  \\\n",
       "6            10     48642  .         C    [T, <*>]       0      .   \n",
       "31           10     48667  .         A    [G, <*>]       0      .   \n",
       "72           10   3642986  .         T    [C, <*>]       0      .   \n",
       "93           10   3643007  .         G    [C, <*>]       0      .   \n",
       "310          10  19491103  .         T    [G, <*>]       0      .   \n",
       "...         ...       ... ..       ...         ...     ...    ...   \n",
       "6960         18  78994238  .         C    [T, <*>]       0      .   \n",
       "6963         18  78994241  .         C    [T, <*>]       0      .   \n",
       "6970         18  78994248  .         G    [A, <*>]       0      .   \n",
       "6972         18  78994250  .         C    [G, <*>]       0      .   \n",
       "7002         18  79891262  .         G    [C, <*>]       0      .   \n",
       "\n",
       "                                                   Info Format        bam file  \n",
       "6     DP=2;I16=0,0,1,0,0,0,70,4900,0,0,60,3600,0,0,6...     PL  60,3,0,60,3,60  \n",
       "31    DP=2;I16=0,0,1,0,0,0,77,5929,0,0,60,3600,0,0,2...     PL  60,3,0,60,3,60  \n",
       "72    DP=2;I16=0,0,1,0,0,0,77,5929,0,0,7,49,0,0,25,6...     PL     7,3,0,7,3,7  \n",
       "93    DP=2;I16=0,0,1,0,0,0,70,4900,0,0,7,49,0,0,25,6...     PL     7,3,0,7,3,7  \n",
       "310   DP=1;I16=0,0,1,0,0,0,40,1600,0,0,60,3600,0,0,1...     PL  40,3,0,40,3,40  \n",
       "...                                                 ...    ...             ...  \n",
       "6960  DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,2...     PL  40,3,0,40,3,40  \n",
       "6963  DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,2...     PL  40,3,0,40,3,40  \n",
       "6970  DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,1...     PL  40,3,0,40,3,40  \n",
       "6972  DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,1...     PL  40,3,0,40,3,40  \n",
       "7002  DP=2;I16=0,0,1,0,0,0,80,6400,0,0,28,784,0,0,25...     PL  28,3,0,28,3,28  \n",
       "\n",
       "[159 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_SNP(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>ID</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>Info</th>\n",
       "      <th>Format</th>\n",
       "      <th>bam file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>10</td>\n",
       "      <td>3643016</td>\n",
       "      <td>.</td>\n",
       "      <td>GCTC</td>\n",
       "      <td>[G]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>INDEL;IDV=2;IMF=1;DP=2;I16=0,0,1,1,0,0,92,5800...</td>\n",
       "      <td>PL</td>\n",
       "      <td>14,6,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2591</th>\n",
       "      <td>10</td>\n",
       "      <td>80247715</td>\n",
       "      <td>.</td>\n",
       "      <td>CGG</td>\n",
       "      <td>[CG]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>INDEL;IDV=1;IMF=1;DP=1;I16=0,0,0,1,0,0,16,256,...</td>\n",
       "      <td>PL</td>\n",
       "      <td>4,3,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7006</th>\n",
       "      <td>18</td>\n",
       "      <td>79891265</td>\n",
       "      <td>.</td>\n",
       "      <td>CT</td>\n",
       "      <td>[CTT]</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>INDEL;IDV=2;IMF=1;DP=2;I16=0,1,1,0,10,100,40,1...</td>\n",
       "      <td>PL</td>\n",
       "      <td>23,1,0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome  Position ID Reference Alternative Quality Filter  \\\n",
       "103          10   3643016  .      GCTC         [G]       0      .   \n",
       "2591         10  80247715  .       CGG        [CG]       0      .   \n",
       "7006         18  79891265  .        CT       [CTT]       0      .   \n",
       "\n",
       "                                                   Info Format bam file  \n",
       "103   INDEL;IDV=2;IMF=1;DP=2;I16=0,0,1,1,0,0,92,5800...     PL   14,6,0  \n",
       "2591  INDEL;IDV=1;IMF=1;DP=1;I16=0,0,0,1,0,0,16,256,...     PL    4,3,0  \n",
       "7006  INDEL;IDV=2;IMF=1;DP=2;I16=0,1,1,0,10,100,40,1...     PL   23,1,0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_INDEL(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_vcf(filename):\n",
    "    \n",
    "    file = list()\n",
    "    \n",
    "    with open(filename, \"r\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\") == False:\n",
    "                row    = [x.replace('\\n', '') for x in line.split('\\t')]\n",
    "                row[4] = row[4].split(\",\")\n",
    "                file.append(row)\n",
    "                \n",
    "    return file\n",
    "\n",
    "def find_SNP(df):\n",
    "    return df[ \n",
    "        df['Reference'].apply(lambda x: len(x)==1) &\n",
    "            (\n",
    "                df['Alternative'].apply(lambda x: len(x)==1 and \"<*>\" not in x and len(x[0]) == 1) |\n",
    "                df['Alternative'].apply(lambda x: len(x)==2 and \"<*>\" in x and len(x[0]) == 1) \n",
    "            )]\n",
    "\n",
    "def find_INDEL(df):\n",
    "    return df[\n",
    "        df.apply(lambda x: len(x['Reference']) != len(x['Alternative'][0]) and x['Alternative'][0] != \"<*>\", axis=1)\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assignment 2: Raw read count for protein coding genes\n",
    "\n",
    "With this assignment we want to calculate how many reads of our SAM file have been mapped on the protein coding genes of chromosome 10 and 18. First of all, we have to obtain names and genomic positions of each protein coding gene from chr10 and chr18.\n",
    "* Write a python program that parses the gtf file and extract information about chromosome 10 and 18 (means first column equal to 10 or 18). Select only rows for which the feature is equal to “gene” and gene_biotype equal to “protein_coding”. Alternatively, you can perform this step using awk bash command (see awk documentation online https://www.gnu.org/software/gawk/manual/gawk.html#Very-Simple).\n",
    "* Use samtools view to filter unmapped reads and supplementary alignments from the SAM file you obtained in LAB2. Save your results into a new SAM file (e.g. unique_aligned.sam)\n",
    "* Using information from the reduced SAM and gtf files, write a Python script to calculate the raw read count for each protein coding gene in chr10 and chr18 (since the original files we provided you were truncated for computational feasibility, you will find a lot of genes with read count equal to zero)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attributes</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1562179</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46892</td>\n",
       "      <td>74163</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000261456\"; gene_version \"5\"; g...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562267</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>134465</td>\n",
       "      <td>254637</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000015171\"; gene_version \"19\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562732</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>274190</td>\n",
       "      <td>689668</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000151240\"; gene_version \"16\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563015</th>\n",
       "      <td>10</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>649948</td>\n",
       "      <td>669581</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000180525\"; gene_version \"11\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563045</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>806914</td>\n",
       "      <td>931705</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000107929\"; gene_version \"14\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410399</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>79970811</td>\n",
       "      <td>80033949</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000141759\"; gene_version \"14\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410520</th>\n",
       "      <td>18</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>80034346</td>\n",
       "      <td>80097088</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000267127\"; gene_version \"7\"; g...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410565</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>80034358</td>\n",
       "      <td>80046397</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000101546\"; gene_version \"12\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410636</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>80109031</td>\n",
       "      <td>80147523</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000101544\"; gene_version \"8\"; g...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410709</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>80157232</td>\n",
       "      <td>80247546</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000178184\"; gene_version \"15\"; ...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>998 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sequence          Source Feature     Start       End Score Strand  \\\n",
       "1562179       10  ensembl_havana    gene     46892     74163     .      -   \n",
       "1562267       10  ensembl_havana    gene    134465    254637     .      +   \n",
       "1562732       10  ensembl_havana    gene    274190    689668     .      -   \n",
       "1563015       10          havana    gene    649948    669581     .      +   \n",
       "1563045       10  ensembl_havana    gene    806914    931705     .      -   \n",
       "...          ...             ...     ...       ...       ...   ...    ...   \n",
       "2410399       18  ensembl_havana    gene  79970811  80033949     .      -   \n",
       "2410520       18          havana    gene  80034346  80097088     .      +   \n",
       "2410565       18  ensembl_havana    gene  80034358  80046397     .      +   \n",
       "2410636       18  ensembl_havana    gene  80109031  80147523     .      +   \n",
       "2410709       18  ensembl_havana    gene  80157232  80247546     .      -   \n",
       "\n",
       "        Phase                                         Attributes  Count  \n",
       "1562179     .  gene_id \"ENSG00000261456\"; gene_version \"5\"; g...      0  \n",
       "1562267     .  gene_id \"ENSG00000015171\"; gene_version \"19\"; ...      0  \n",
       "1562732     .  gene_id \"ENSG00000151240\"; gene_version \"16\"; ...      0  \n",
       "1563015     .  gene_id \"ENSG00000180525\"; gene_version \"11\"; ...      0  \n",
       "1563045     .  gene_id \"ENSG00000107929\"; gene_version \"14\"; ...      0  \n",
       "...       ...                                                ...    ...  \n",
       "2410399     .  gene_id \"ENSG00000141759\"; gene_version \"14\"; ...      0  \n",
       "2410520     .  gene_id \"ENSG00000267127\"; gene_version \"7\"; g...      0  \n",
       "2410565     .  gene_id \"ENSG00000101546\"; gene_version \"12\"; ...      0  \n",
       "2410636     .  gene_id \"ENSG00000101544\"; gene_version \"8\"; g...      0  \n",
       "2410709     .  gene_id \"ENSG00000178184\"; gene_version \"15\"; ...      0  \n",
       "\n",
       "[998 rows x 10 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GTF_FILE = \"Homo_sapiens.GRCh38.95.gtf\"\n",
    "\n",
    "file = parse_gtf(GTF_FILE)\n",
    "\n",
    "gtf_df = pd.DataFrame(file, columns=['Sequence', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Phase', 'Attributes'])\n",
    "\n",
    "gtf_df[\"Count\"] = 0\n",
    "\n",
    "gtf_df = gtf_df.query('Sequence == \"10\" or Sequence == \"18\"')\n",
    "gtf_df = extract_protein_coding_genes(gtf_df)\n",
    "gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QNAME</th>\n",
       "      <th>FLAG</th>\n",
       "      <th>RNAME</th>\n",
       "      <th>POS</th>\n",
       "      <th>MAPQ</th>\n",
       "      <th>CIGAR</th>\n",
       "      <th>RNEXT</th>\n",
       "      <th>PNEXT</th>\n",
       "      <th>TLEN</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>LENGTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX288614.1</td>\n",
       "      <td>99</td>\n",
       "      <td>10</td>\n",
       "      <td>55667614</td>\n",
       "      <td>39</td>\n",
       "      <td>12S79M1I9M</td>\n",
       "      <td>=</td>\n",
       "      <td>55667850</td>\n",
       "      <td>318</td>\n",
       "      <td>TTTTTCTAGACGGCAGGTCAGGTCCACCACTGACACGTTGGCAGTG...</td>\n",
       "      <td>&lt;&lt;BFFFF0BBF&lt;0FFFFIBB0BFBF7BF0&lt;&lt;BFB0BFBF&lt;&lt;&lt;0BFB...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX288614.1</td>\n",
       "      <td>147</td>\n",
       "      <td>10</td>\n",
       "      <td>55667850</td>\n",
       "      <td>39</td>\n",
       "      <td>19S82M</td>\n",
       "      <td>=</td>\n",
       "      <td>55667614</td>\n",
       "      <td>-318</td>\n",
       "      <td>AGTCCTTCCACGATACCAAAGTTGTCATGGATGTCCTTGGCCAGGG...</td>\n",
       "      <td>B&lt;&lt;&lt;BB&lt;&lt;&lt;&lt;&lt;B&lt;B&lt;B&lt;&lt;&lt;B&lt;B&lt;BBBB&lt;&lt;&lt;&lt;77'&lt;7&lt;BBBBBBBBB...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERX288614.2</td>\n",
       "      <td>185</td>\n",
       "      <td>10</td>\n",
       "      <td>45783567</td>\n",
       "      <td>0</td>\n",
       "      <td>38M63S</td>\n",
       "      <td>=</td>\n",
       "      <td>45783567</td>\n",
       "      <td>0</td>\n",
       "      <td>CCTCCCACCTCAGCCTCCCAAGTAGCTGGAACCACAGGAGGGTAAC...</td>\n",
       "      <td>&lt;&lt;0&lt;B77BBBBBBB&lt;7BBBBBB&lt;BBBFBBBBB&lt;&lt;FFFFFIIFB&lt;B&lt;...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERX288614.8</td>\n",
       "      <td>185</td>\n",
       "      <td>18</td>\n",
       "      <td>31416709</td>\n",
       "      <td>0</td>\n",
       "      <td>46S55M</td>\n",
       "      <td>=</td>\n",
       "      <td>31416709</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAGGACATTGTGGTTAATTTTTTTTAATTTTTGGTTTTTTTGGGG...</td>\n",
       "      <td>#####################################BB777'0'0...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERX288614.11</td>\n",
       "      <td>83</td>\n",
       "      <td>10</td>\n",
       "      <td>100057456</td>\n",
       "      <td>60</td>\n",
       "      <td>101M</td>\n",
       "      <td>=</td>\n",
       "      <td>100057454</td>\n",
       "      <td>-103</td>\n",
       "      <td>GTATAAGAGATCAGGTTCGTCCTTTAGTGTTGTGTATGGTTATCAT...</td>\n",
       "      <td>BBBBBFBFBBBBBBBBBBBBBBBBB&lt;BBBBBFFFFFFFFFFFFIII...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>ERX288614.390</td>\n",
       "      <td>163</td>\n",
       "      <td>10</td>\n",
       "      <td>21397719</td>\n",
       "      <td>40</td>\n",
       "      <td>24S19M58S</td>\n",
       "      <td>=</td>\n",
       "      <td>21397719</td>\n",
       "      <td>19</td>\n",
       "      <td>TTTCATTTGCTGGTGTGAGCCATGGGGTGTGTATTTGTATGTGGGG...</td>\n",
       "      <td>BBBFFFFFFFFFBFFBBFFFFIBFFFI'B&lt;&lt;&lt;&lt;BFFFFBBFBFFFF...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>ERX288614.393</td>\n",
       "      <td>83</td>\n",
       "      <td>10</td>\n",
       "      <td>36523493</td>\n",
       "      <td>60</td>\n",
       "      <td>101M</td>\n",
       "      <td>=</td>\n",
       "      <td>36523405</td>\n",
       "      <td>-189</td>\n",
       "      <td>TTACAATATGTTCAAAAGCATCTTTTTCATGGTCTTTCCCCCAAGC...</td>\n",
       "      <td>###########BBBBBB&lt;BB&lt;BBBBB&lt;BBB7BB&lt;7B0FFBBB&lt;&lt;FB...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>ERX288614.393</td>\n",
       "      <td>163</td>\n",
       "      <td>10</td>\n",
       "      <td>36523405</td>\n",
       "      <td>60</td>\n",
       "      <td>101M</td>\n",
       "      <td>=</td>\n",
       "      <td>36523493</td>\n",
       "      <td>189</td>\n",
       "      <td>CTTAGATCTTCACCCCATATTTTCTCACACGCATTATAAATGTCAT...</td>\n",
       "      <td>B0&lt;BBFBFFBFBFBFFBB&lt;FFFBFFFIFBFB7BFFB&lt;FBFFFFBB&lt;...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>ERX288614.399</td>\n",
       "      <td>83</td>\n",
       "      <td>10</td>\n",
       "      <td>98800042</td>\n",
       "      <td>0</td>\n",
       "      <td>20M81S</td>\n",
       "      <td>=</td>\n",
       "      <td>98800043</td>\n",
       "      <td>-19</td>\n",
       "      <td>GAACAAAGAAAAAAAAAGGGGAGGGAAAGGACACCAAGTCTTGGGG...</td>\n",
       "      <td>0BBBFFBBFFFFFBBBFFFFFFFBBBFFFBBBBFFFFFFFFFFIII...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>ERX288614.399</td>\n",
       "      <td>163</td>\n",
       "      <td>10</td>\n",
       "      <td>98800043</td>\n",
       "      <td>0</td>\n",
       "      <td>4S19M78S</td>\n",
       "      <td>=</td>\n",
       "      <td>98800042</td>\n",
       "      <td>19</td>\n",
       "      <td>CAGGAACAAAGAAAAAAAAAGGGGAGGGAAAGGACACCAAGTCTTG...</td>\n",
       "      <td>BBBFFFFFFFFFFIIIIIIIIIIIFFIIIIIIFBFBFFFFFFFFFF...</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>227 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             QNAME FLAG RNAME        POS MAPQ       CIGAR RNEXT      PNEXT  \\\n",
       "0      ERX288614.1   99    10   55667614   39  12S79M1I9M     =   55667850   \n",
       "1      ERX288614.1  147    10   55667850   39      19S82M     =   55667614   \n",
       "2      ERX288614.2  185    10   45783567    0      38M63S     =   45783567   \n",
       "3      ERX288614.8  185    18   31416709    0      46S55M     =   31416709   \n",
       "4     ERX288614.11   83    10  100057456   60        101M     =  100057454   \n",
       "..             ...  ...   ...        ...  ...         ...   ...        ...   \n",
       "222  ERX288614.390  163    10   21397719   40   24S19M58S     =   21397719   \n",
       "223  ERX288614.393   83    10   36523493   60        101M     =   36523405   \n",
       "224  ERX288614.393  163    10   36523405   60        101M     =   36523493   \n",
       "225  ERX288614.399   83    10   98800042    0      20M81S     =   98800043   \n",
       "226  ERX288614.399  163    10   98800043    0    4S19M78S     =   98800042   \n",
       "\n",
       "     TLEN                                                SEQ  \\\n",
       "0     318  TTTTTCTAGACGGCAGGTCAGGTCCACCACTGACACGTTGGCAGTG...   \n",
       "1    -318  AGTCCTTCCACGATACCAAAGTTGTCATGGATGTCCTTGGCCAGGG...   \n",
       "2       0  CCTCCCACCTCAGCCTCCCAAGTAGCTGGAACCACAGGAGGGTAAC...   \n",
       "3       0  AAAGGACATTGTGGTTAATTTTTTTTAATTTTTGGTTTTTTTGGGG...   \n",
       "4    -103  GTATAAGAGATCAGGTTCGTCCTTTAGTGTTGTGTATGGTTATCAT...   \n",
       "..    ...                                                ...   \n",
       "222    19  TTTCATTTGCTGGTGTGAGCCATGGGGTGTGTATTTGTATGTGGGG...   \n",
       "223  -189  TTACAATATGTTCAAAAGCATCTTTTTCATGGTCTTTCCCCCAAGC...   \n",
       "224   189  CTTAGATCTTCACCCCATATTTTCTCACACGCATTATAAATGTCAT...   \n",
       "225   -19  GAACAAAGAAAAAAAAAGGGGAGGGAAAGGACACCAAGTCTTGGGG...   \n",
       "226    19  CAGGAACAAAGAAAAAAAAAGGGGAGGGAAAGGACACCAAGTCTTG...   \n",
       "\n",
       "                                                  QUAL  LENGTH  \n",
       "0    <<BFFFF0BBF<0FFFFIBB0BFBF7BF0<<BFB0BFBF<<<0BFB...     101  \n",
       "1    B<<<BB<<<<<B<B<B<<<B<B<BBBB<<<<77'<7<BBBBBBBBB...     101  \n",
       "2    <<0<B77BBBBBBB<7BBBBBB<BBBFBBBBB<<FFFFFIIFB<B<...     101  \n",
       "3    #####################################BB777'0'0...     101  \n",
       "4    BBBBBFBFBBBBBBBBBBBBBBBBB<BBBBBFFFFFFFFFFFFIII...     101  \n",
       "..                                                 ...     ...  \n",
       "222  BBBFFFFFFFFFBFFBBFFFFIBFFFI'B<<<<BFFFFBBFBFFFF...     101  \n",
       "223  ###########BBBBBB<BB<BBBBB<BBB7BB<7B0FFBBB<<FB...     101  \n",
       "224  B0<BBFBFFBFBFBFFBB<FFFBFFFIFBFB7BFFB<FBFFFFBB<...     101  \n",
       "225  0BBBFFBBFFFFFBBBFFFFFFFBBBFFFBBBBFFFFFFFFFFIII...     101  \n",
       "226  BBBFFFFFFFFFFIIIIIIIIIIIFFIIIIIIFBFBFFFFFFFFFF...     101  \n",
       "\n",
       "[227 rows x 12 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAM_FILE = \"unique_aligned.sam\"\n",
    "\n",
    "file   = parse_SAM(SAM_FILE)\n",
    "sam_df = pd.DataFrame(file, columns=['QNAME', 'FLAG', 'RNAME', 'POS', 'MAPQ', 'CIGAR', 'RNEXT', 'PNEXT', 'TLEN', 'SEQ', 'QUAL'])\n",
    "\n",
    "sam_df[\"LENGTH\"] = sam_df.apply(lambda x: len(x.SEQ), axis=1)\n",
    "\n",
    "sam_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attributes</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1562179</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46892</td>\n",
       "      <td>74163</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000261456\"; gene_version \"5\"; g...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562267</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>134465</td>\n",
       "      <td>254637</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000015171\"; gene_version \"19\"; ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563589</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1177318</td>\n",
       "      <td>1737476</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185736\"; gene_version \"15\"; ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569992</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>12349482</td>\n",
       "      <td>12835545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183049\"; gene_version \"12\"; ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1570132</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>13099449</td>\n",
       "      <td>13138308</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000123240\"; gene_version \"16\"; ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405878</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>70003031</td>\n",
       "      <td>70205945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000176225\"; gene_version \"13\"; ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2406784</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>72536680</td>\n",
       "      <td>72638521</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000141668\"; gene_version \"9\"; g...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408563</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>76978827</td>\n",
       "      <td>77133683</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000197971\"; gene_version \"14\"; ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2409234</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>78980275</td>\n",
       "      <td>79002677</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256463\"; gene_version \"8\"; g...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410189</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>79863668</td>\n",
       "      <td>79900184</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000178342\"; gene_version \"4\"; g...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sequence          Source Feature     Start       End Score Strand  \\\n",
       "1562179       10  ensembl_havana    gene     46892     74163     .      -   \n",
       "1562267       10  ensembl_havana    gene    134465    254637     .      +   \n",
       "1563589       10  ensembl_havana    gene   1177318   1737476     .      -   \n",
       "1569992       10  ensembl_havana    gene  12349482  12835545     .      +   \n",
       "1570132       10  ensembl_havana    gene  13099449  13138308     .      +   \n",
       "...          ...             ...     ...       ...       ...   ...    ...   \n",
       "2405878       18  ensembl_havana    gene  70003031  70205945     .      -   \n",
       "2406784       18  ensembl_havana    gene  72536680  72638521     .      -   \n",
       "2408563       18  ensembl_havana    gene  76978827  77133683     .      -   \n",
       "2409234       18  ensembl_havana    gene  78980275  79002677     .      +   \n",
       "2410189       18  ensembl_havana    gene  79863668  79900184     .      +   \n",
       "\n",
       "        Phase                                         Attributes  Count  \n",
       "1562179     .  gene_id \"ENSG00000261456\"; gene_version \"5\"; g...      2  \n",
       "1562267     .  gene_id \"ENSG00000015171\"; gene_version \"19\"; ...      2  \n",
       "1563589     .  gene_id \"ENSG00000185736\"; gene_version \"15\"; ...      2  \n",
       "1569992     .  gene_id \"ENSG00000183049\"; gene_version \"12\"; ...      1  \n",
       "1570132     .  gene_id \"ENSG00000123240\"; gene_version \"16\"; ...      2  \n",
       "...       ...                                                ...    ...  \n",
       "2405878     .  gene_id \"ENSG00000176225\"; gene_version \"13\"; ...      1  \n",
       "2406784     .  gene_id \"ENSG00000141668\"; gene_version \"9\"; g...      2  \n",
       "2408563     .  gene_id \"ENSG00000197971\"; gene_version \"14\"; ...      2  \n",
       "2409234     .  gene_id \"ENSG00000256463\"; gene_version \"8\"; g...      2  \n",
       "2410189     .  gene_id \"ENSG00000178342\"; gene_version \"4\"; g...      2  \n",
       "\n",
       "[101 rows x 10 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perform_raw_read_count(gtf_df, sam_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gtf(filename):\n",
    "    \n",
    "    file = list()\n",
    "    \n",
    "    with open(filename, \"r\") as f:\n",
    "        for line in f: \n",
    "            if line.startswith(\"#\") == False:\n",
    "                row = [x.replace('\\n', '') for x in line.split('\\t')]\n",
    "                file.append(row)\n",
    "        \n",
    "    return file\n",
    "\n",
    "def parse_SAM(filename):\n",
    "    \n",
    "    file = list()\n",
    "    \n",
    "    with open(filename, \"r\") as f:\n",
    "        for line in f: \n",
    "            if line.startswith(\"#\") == False:\n",
    "                row = [x.replace('\\n', '') for x in line.split('\\t')]\n",
    "                file.append(row[:11])\n",
    "        \n",
    "    return file\n",
    "\n",
    "def extract_protein_coding_genes(df):\n",
    "    return df[df.apply(lambda x: x.Feature == \"gene\" and \"protein_coding\" in x.Attributes, axis=1)]\n",
    "\n",
    "def perform_raw_read_count(gtf_df, sam_df):\n",
    "    \n",
    "    for gtf_index, gtf_row in gtf_df.iterrows():\n",
    "        \n",
    "        for sam_index, sam_row in sam_df.iterrows():\n",
    "            \n",
    "            if int(gtf_row.Start) <= int(sam_row.POS) and int(sam_row.POS) + sam_row.LENGTH <= int(gtf_row.End):\n",
    "                gtf_df.loc[gtf_index, 'Count'] += 1\n",
    "            \n",
    "    return gtf_df[gtf_df.Count != 0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
